README.md
crawler-mswl: Python web crawler
=================================

Introduction
=============

This is a Python crawler project to retrieve all links in `a` tags defined in URL and access every `href` property for each link as defined in input parameter.

Ricardo García Fernández developed this project for ``Developing Tools`` Subject coursed in [Master on Libre Software (Master Universitario en Software libre](http://master.libresoft.es/) at [Universidad Rey juan Carlos](http://www.urjc.es/).

Requirements
=============

* Java JDK 1.6.0_22
* Python 2.7.2

License
========

This project has ``GNU GENERAL PUBLIC LICENSE version 3`` see complete definition in [COPYING](COPYING) file.

Install
========

Read install instructions in [INSTALL](INSTALL) file.

* Install from PYPI.
* Install from Source.

Command
========

How to use the application.

This application run by command `ricardo-crawler` after instalation.

To use this application without installing it you have to type:

```
#!shell
python ricardo-crawler.py -n 1 http://www.google.es

```

Arguments
----------

Two arguments to run the application:
```
#!shell
-n [Number of levels to inspect]: number of depth level to retrieve 
  links. If no argument is passed, by default is assigned to *1*.
[URL]: URL to inspect for each level defined.
```

Has a help command to define correctly these arguments.
```
#!shell
python ricardo_crawler.py --help
```

Help Console Output:

```
#!shell
usage: ricardo_crawler.py [-h] [-n NUMBER_OF_LEVELS] url

Let's crawl a web

positional arguments:
  url                   target URL

optional arguments:
  -h, --help            show this help message and exit
  -n NUMBER_OF_LEVELS, --number-of-levels NUMBER_OF_LEVELS
                        how depth the crawl will go.
```

Test
-----

There are tests developer for each controller behaviour:

```
test/
├── controllers
│   ├── __init__.py
│   ├── __init__.pyc
│   ├── TestHttpLinksCollector.py
│   ├── TestHttpLinksCollector.pyc
│   ├── TestUrlContentRetrieve.py
│   └── TestUrlContentRetrieve.pyc
├── __init__.py
└── __init__.pyc
```

Issues management
==================

Issue management its done with Redmine project tracker in:

  http://code.sidelab.es/projects/ricardogarcia2012/issues

License
========

<a href="http://creativecommons.org/licenses/by/3.0/" rel="Creative Commons Attribution 3.0">![Foo](http://i.creativecommons.org/l/by/3.0/88x31.png)</a>

This work is licensed under a [Creative Commons Attribution 3.0 Unported License](http://creativecommons.org/licenses/by/3.0/).


Changelog
=================================
crawler-mswl: Python web crawler
=================================

Version 00.09.20
=================

Install version using ``easy_install ricardo-crawler``.
Input argument control.
Exception related to URL controlled and wrote in log file in ``/var/tmp/crawler.log``.
Result output in log file.

Version 1.00.00
=================

Documentation complete (TBC).
Pylint Test passed.
Test sources removed from install version.


test/controllers/TestHttpLinksCollector.py
# coding=UTF-8
'''
    This file is part of crawler-mswl.

    crawler-mswl is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    crawler-mswl is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with crawler-mswl.  If not, see <http://www.gnu.org/licenses/>.

'''

import unittest
from controllers.http_links_collector import HttpLinksCollector

class TestHttpLinksCollector (unittest.TestCase):
    '''Test class for methods defined in HttpLinksCollector class.
    
    Created on 01/10/2012

    @author: Ricardo García Fernández
    @mail: ricardogarfe@gmail.com

    '''

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_retrieveLinks(self):
        '''Test method for HttpLinksCollector.retrieve_links.
        
        '''
        
        # Test 1 - check HttpError 401 in log
        starting_url = "http://www.nature.com"
        target_url = "http://www.nature.com/nature/journal/v438/n7070/full/438900a.html"

        http_links_collector = HttpLinksCollector(starting_url)        
        links_retrieved = http_links_collector.retrieve_links(target_url)
        
        self.assertTrue(not links_retrieved, \
                       "Retreved Links from:'" + target_url + "'")
        
        # Test 2 - check URLError - protocol irc.
        starting_url = "http://www.nature.com"
        target_url = "irc://irc.freenode.net/wikimedia-ayuda"

        http_links_collector = HttpLinksCollector(starting_url)        
        links_retrieved = http_links_collector.retrieve_links(target_url)
        
        self.assertTrue(not links_retrieved, \
                       "Retreved Links from:'" + target_url + "'")

if __name__ == "__main__":
    #import sys;sys.argv = ['', 'Test.testRetrieveLinks']
    unittest.main()


test/controllers/TestUrlContentRetrieve.py
'''
Created on 01/10/2012

@author: ricardo
'''
import unittest
from controllers.url_content_retrieve import UrlContentRetrieve
from urllib2 import URLError

class TestUrlContentRetrieve(unittest.TestCase):
    '''Test class for methods defined in UrlContentRetrieve class.
    '''

    def setUp(self):
        pass


    def tearDown(self):
        pass


    def test_generate_correct_url(self):
        ''' Test method for retrieve the correct URL
        '''
        
        starting_url = "http://localhost"
        url_content_retrieve = UrlContentRetrieve(starting_url)

        # Test 1
        url_to_check = "http://localhost"
        expected_url = "http://localhost"
        returned_url = url_content_retrieve.generate_correct_url(url_to_check)
        
        self.assertEqual(returned_url, expected_url, \
                         "Error converting URL\n:\t* from: "\
                          + url_to_check + "\n\t* to: " + expected_url + \
                          "\nReturned value: " + returned_url)

        # Test 2    
        url_to_check = "/media"
        expected_url = "http://localhost/media"
        returned_url = url_content_retrieve.generate_correct_url(url_to_check)
        
        self.assertEqual(returned_url, expected_url, \
                         "Error converting URL\n:\t* from: "\
                          + url_to_check + "\n\t* to: " + expected_url + \
                          "\nReturned value: " + returned_url)

        # Test 3
        url_to_check = "//mediateca.org"
        expected_url = "http://mediateca.org"
        returned_url = url_content_retrieve.generate_correct_url(url_to_check)
        
        self.assertEqual(returned_url, expected_url, \
                         "Error converting URL\n:\t* from: "\
                          + url_to_check + "\n\t* to: " + expected_url + \
                          "\nReturned value: " + returned_url)
        
    def test_url_content(self):
        '''Test method for retrieve content from url.
        '''
        
        # Initialize object
        starting_url = "http://non-existing-url.com"
        url_content_retrieve = UrlContentRetrieve(starting_url)

        # Test 1 - Variables
        # Return error
        target_url = "http://non-existing-url.com"
        expected_exception = \
            "<urlopen error [Errno -2] Name or service not known>"
        try:
            returned_soup_code = url_content_retrieve.url_content(target_url)
            self.assertIsNone(returned_soup_code, \
                              "Url content has return a value:\n"\
                              + str(returned_soup_code))
        except URLError, url_error:
            self.assertEqual(expected_exception, str(url_error), "Exception "\
                             "URLError is not correct with message:\n"\
                             + str(url_error))
        
        # Test 2 - enconding
        starting_url = "http://es.wikipedia.org"
        url_content_retrieve = UrlContentRetrieve(starting_url)
        target_url = \
            "http://es.wikipedia.org/wiki/Depresi%C3%B3n_tropical_Diez_(2007)"
        expected_enconding = "utf-8"
        
        returned_soup_code = url_content_retrieve.url_content(target_url)
        
        returned_enconding = returned_soup_code.originalEncoding

        self.assertEqual(expected_enconding, returned_enconding, \
                         "Encondin doesn't match:\n\t* Expected: "\
                         + expected_enconding + "\n\t* Returned: "\
                         + returned_enconding)
            

if __name__ == "__main__":
    #import sys;sys.argv = ['', 'Test.testName']
    unittest.main()


ricardo-crawler.py
'''
Created on 27/09/2012

@author: ricardo
'''

import urllib2
from setuptools.package_index import user_agent
from urllib2 import _opener
from BeautifulSoup import BeautifulSoup as Soup
import argparse

user_agent = "Mozilla/5.0 (X11; U; Linux x86_64; en-Us) AppleWebKit/534.7 (KHTML, Like Gecko) Chrome /7.0.517.41 Safari/534.7"

_opener = urllib2.build_opener()
_opener.addheaders = [('User-agent', user_agent)]

# Parameters
parser = argparse.ArgumentParser(description="Let's crawl a web")

parser.add_argument('url', nargs=1, help='target URL')
parser.add_argument('-n', '--number-of-levels', type=int, default=1, help='how depth the crawl will go.')

args = parser.parse_args()

target_url = args.url.pop()
depth = args.number_of_levels
print depth

# URL Decode
raw_code = _opener.open(target_url).read()

soup_code = Soup(raw_code)
links = []

for link in soup_code.findAll('a') :
         if link.has_key('href'):
            links.append(link['href'])
            print link['href']


ricardo_crawler.py
#!/usr/bin/env python
# encoding=utf-8
'''
    This file is part of crawler-mswl.

    crawler-mswl is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    crawler-mswl is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with crawler-mswl.  If not, see <http://www.gnu.org/licenses/>.

'''

import argparse
from controllers.http_links_collector import HttpLinksCollector

class RicardoCrawler:
    '''RicardoCrawler class to crawl a web using system parameters.

    Created on 27/09/2012

    @author: Ricardo García Fernández
    @mail: ricardogarfe@gmail.com
    
    '''
    
    def __init__(self):
        '''Init method
        '''
        
        self.crawler_start()

    def crawler_start(self):
        '''Method to start crawling.
            * Checks input parameters.
            * returns the result of crawling printing a dictionary on the screen.
        
        '''

        # ArgParse definition rules
        parser = argparse.ArgumentParser(description="Let's crawl a web")
        
        parser.add_argument('url', nargs=1, help='target URL')
        parser.add_argument('-n', '--number-of-levels', type = int, \
                            default = 1, help = 'how depth the crawl will go.')
        
        # Create argument object
        args = parser.parse_args()
        
        target_url = args.url.pop()
        depth = args.number_of_levels
        
        # Starting level to retrieve links
        level = 1
        links = {}
        http_links_collector = HttpLinksCollector(target_url)
        
        links_list = http_links_collector.\
            retrieve_links(target_url, depth, level)
        
        links[target_url] = links_list
        
        print links


def main():
    RicardoCrawler()

if __name__ == '__main__':
    main()     


controllers/http_links_collector.py
# coding=UTF-8
'''
    This file is part of crawler-mswl.

    crawler-mswl is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    crawler-mswl is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with crawler-mswl.  If not, see <http://www.gnu.org/licenses/>.

'''

from url_content_retrieve import UrlContentRetrieve
import logging

class HttpLinksCollector:
    '''Class to manage links from url.
    
    Created on 27/09/2012

    @author: Ricardo García Fernández
    @mail: ricardogarfe@gmail.com

    '''

    def __init__(self, starting_url):
        '''Initialize to set urlContentRetrieve object with startingURL.
        
        Keyword arguments:
        starting_url -- URL to start crawling.

        '''
        
        # Setup Log
        self.setup_log()
        # Define url content retrieve to use
        self.url_content_retrieve = UrlContentRetrieve(starting_url)
    
    def retrieve_links(self, target_url, depth=1, level=1):
        '''
        Retrieve links from url content until defined depth organized in levels.
        
        Keyword arguments:
        target_url -- URL to analyze content and retrive links.
        depth -- Depth of links to analyze.
        level -- Level in which start to analyze.
        
        '''

        links = {}

        if depth >= level:

            soup_code = self.url_content_retrieve.url_content(target_url)
            
            if soup_code:
                formatted_links = \
                    self.url_content_retrieve.\
                        retrieve_formatted_links(soup_code)
                for link in formatted_links :
    
                        self.logger.info(self.print_depth(level) + " " + link)
    
                        try:
                            sublinks = \
                                self.retrieve_links(link, depth, level + 1)
                            links[link] = sublinks
                        except ValueError, value_error:
                            # Invalid URL
                            self.logger.error("URL is not correct:\t" + link + \
                                              "\nException:\t"\
                                               + str(value_error))
        return links
    
    def print_depth(self, level):
        '''
        Creates the string with the '*' character serie appropriate to the deep.

        Keyword arguments:
        level -- level to print *
        '''

        deepLetter = ""
        for i in range(level):
            deepLetter = deepLetter + "*"
            
        return deepLetter

    def setup_log(self):
        '''Set up Python logging.
        
        '''

        self.logger = logging.getLogger('HttpLinksCollector')
        self.hdlr = logging.FileHandler('/var/tmp/HttpLinksCollector.log')
        self.formatter = logging.Formatter('%(asctime)s %(levelname)s %(message)s')
        self.hdlr.setFormatter(self.formatter)
        self.logger.addHandler(self.hdlr)
        self.logger.setLevel(logging.INFO)



controllers/url_content_retrieve.py
# coding=UTF-8
'''
    This file is part of crawler-mswl.

    crawler-mswl is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    crawler-mswl is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with crawler-mswl.  If not, see <http://www.gnu.org/licenses/>.

'''

import urllib2
import logging
from BeautifulSoup import BeautifulSoup as Soup
from urllib2 import HTTPError, URLError
import traceback

class UrlContentRetrieve:
    '''Controller class to manage url and retrieve content using BeautifulSoup.

    Created on 27/09/2012

    @author: Ricardo García Fernández
    @mail: ricardogarfe@gmail.com

    '''

    def __init__(self, top_url):
        '''Initializes UrlContentRetrieve with default user_agent.
        
        Keyword arguments:
        top_url -- URL from top web level.
        
        '''
        
        # Setup Log
        self.setup_log()

        # Define user agent
        user_agent = "Mozilla/5.0 (X11; U; Linux x86_64; en-Us) \
            AppleWebKit/534.7 (KHTML, Like Gecko) Chrome \
            /7.0.517.41 Safari/534.7"

        self.opener = urllib2.build_opener()
        self.opener.addheaders = [('User-agent', user_agent)]
        
        if top_url.endswith("/"):
            self.top_url = top_url[:-1]
        self.top_url = top_url

    def url_content(self, target_url):
        ''' Retrieve URL raw Content.
        
        Keyword arguments:
        target_url -- url to inspect.
        
        '''
        
        # Get correct format of URL
        target_url = self.generate_correct_url(target_url)

        try:
            raw_code = self.opener.open(target_url)
            soup_code = Soup(raw_code, fromEncoding="utf-8")
        except HTTPError, http_error:
            self.logger.error("HttpError with url:\t" + target_url + \
                              "\nException message:\t " + \
                                str(http_error)\
                                + "\nStack trace:\t" + \
                                traceback.format_exc())
            return None
        except URLError, url_error:
            self.logger.error("URLError with url:\t" + target_url + \
                              "\nException message:\t " + \
                                str(url_error)\
                                + "\nStack trace:\t" + \
                                traceback.format_exc())
            return None

        return soup_code

    def generate_correct_url(self, url_to_check):
        ''' Check the url and generate a new one well formed if is not complete.
            * Url management starts with '/' or /wiki/ - InvalidURL
            * Url management starts with '//' - HTTPError
            * Url format t0 utf-8 - http://es.wikipedia.org//yi.wikipedia.org/wiki/%D7%B0%D7%99%D7%A7%D7%99%D7%A4%D6%BC%D7%A2%D7%93%D7%99%D7%A2:%D7%91%D7%A8%D7%95%D7%9B%D7%99%D7%9D_%D7%94%D7%91%D7%90%D7%99%D7%9D
        
        Keyword arguments:
        url_to_check -- URL to check

        '''

        try:
            url_to_check = str(url_to_check.encode('utf-8'))
        except UnicodeEncodeError, unicode_error:
            self.logger.warning("UnicodeEncodeError with url:\t" + \
                              url_to_check + "\nException message:\t " + \
                                str(unicode_error)\
                                + "\nStack trace:\t" + \
                                traceback.format_exc())
            return None

        if url_to_check.startswith("//"):
            url_to_check = "http:" + url_to_check
        elif url_to_check.startswith("/"):
            url_to_check = self.top_url + url_to_check
        elif url_to_check.startswith("#"):
            url_to_check = None
        elif not url_to_check.find("https://") and \
            not url_to_check.find("http://"):
            url_to_check = self.top_url + url_to_check

        return url_to_check

    def retrieve_formatted_links(self, soup_code):
        ''' Retrieve the links in correct format into a list.
        
        '''

        formatted_links = []
        raw_links = []

        # Extract all links
        for link in soup_code.findAll('a') :
            if link.has_key('href'):
                raw_links.append(link['href'])
        
        # Convert to correct format
        for raw_link in raw_links:
            formatted_link = self.generate_correct_url(raw_link)
            if formatted_link:
                formatted_links.append(formatted_link)
            
        return formatted_links

    def to_unicode(self, str, verbose=False):
        '''attempt to fix non uft-8 string into utf-8, using a limited set of encodings'''

        # fuller list of encodings at http://docs.python.org/library/codecs.html#standard-encodings
        if not str:  return u''
        u = None
        # we could add more encodings here, as warranted.  
        encodings = ('ascii', 'utf8', 'latin1')
        for enc in encodings:
            if u:  break
            try:
                u = unicode(str, enc)
            except UnicodeDecodeError:
                if verbose: print "error for %s into encoding %s" % (str, enc)
                pass
        if not u:
            u = unicode(str, errors='replace')
            if verbose:  print "using replacement character for %s" % str
        return u

    def setup_log(self):
        '''Setup Python logging.
        
        '''

        self.logger = logging.getLogger('UrlContentRetrieve')
        self.hdlr = logging.FileHandler('/var/tmp/crawler.log')
        self.formatter = logging.Formatter('%(asctime)s %(levelname)s \
            %(filename)s %(message)s')
        self.hdlr.setFormatter(self.formatter)
        self.logger.addHandler(self.hdlr)
        self.logger.setLevel(logging.WARNING)
